ynput/ayon-core
# Houdini: Publish any ROP node (Generic Creator)

## Changelog Description

This PR implements an idea for "lower level publishing" in Houdini. This implement **Generic ROP publishing**. Just create any Houdini ROP node (or custom Rop node HDA) and publish your product types from it!

> This PR adds a **Generic Creator**:
> Currently, It injects the ayon meta data onto ROP nodes to behave as though they were produced by various creators from Ayon. This ensures that validators and collectors are retained without loss.
> e.g. using Generic creator while setting product type to pointcache should yield similar/same results as using pointcache creator
> Here is an example using labs karma node and make it behave as though it was made by karma creator. 
> https://github.com/ynput/ayon-core/pull/542#issuecomment-2179333623

_This PR originally also contained a Dynamic Runtime Instance creator. That is now separated to another PR here:_ https://github.com/ynput/ayon-core/pull/691

## Additional info

_As part of the Ynput Houdini Workgroup sessions I developed this quick prototype to expose a way to batch publish and ingest files. Consider it more of an exploration of what's possible then a "drop it in production now" ready-to-go solution._

#### Explainer

Yes, this requires some explanation. Here you go.

https://github.com/ynput/ayon-core/assets/2439881/5f767493-10bd-41f6-b54f-89313a96da33

What I forgot to add is that it currently still relies on detecting what the output files are for a ROP node based on a "file" parm that is often unique in Houdini per node. If anyone knows a way to just query the expected output files for a ROP node (similar to what PDG/TOPs seems to do I'd love to know!) but otherwise we'll just need to expand that list.

However, I also played with the idea of having "custom file list" attributes on the Node that when enabled could override the "Collector" logic and would instead use that list of files as the publish instance's files. So that e.g. one instance could also publish multiple representations. For that, @MustafaJafar did this lovely non-functional 'prototype' but it does get the idea across:

![mustafa_collect_multiple_representations_to_one_instance](https://github.com/ynput/ayon-core/assets/2439881/e33fede1-60d3-4307-a881-21bd2858cb76)


#### TODO

- [x] Use `$OS` as default _variant_ name.
- [x] Add "OnCreated" callback to allow a studio to configure which nodes to automatically imprint as 'publishable'. [See Fabia's first video here at 08:10](https://github.com/ynput/ayon-core/pull/542#issuecomment-2128044486) and [code here](https://github.com/fabiaserra/ayon-core/blob/release/alkemyx/server_addon/houdini/client/ayon_houdini/startup/OnCreated.py)
    - [ ] This is currently a first iteration - should still be exposed to project settings and allow user configuration.
- [x] The product types being an enum with predefined standards per node type.
- [ ] Implement good support for multiple representations, e.g. [like this](https://github.com/BigRoy/ayon-core/pull/6) and [Fabia's second video near the end](https://github.com/ynput/ayon-core/pull/542#issuecomment-2128044486).
- [ ] The comment field being available on the ROP nodes
    - [x] This does actually already work if you'd add the product type to `ayon+settings://core/publish/collect_comment_per_instance` (although not houdini-specific nor as nice?)
- [ ] Add "Publish" button directly on the node similar to "self publish" or whatever we had for others
- [ ] Allow exposing to settings which node types may need to be "automatically imprinted" on create for a studio to customize, with also settings for allowed product types (the default type, default variant name, etc.)
- [ ] Make sure [this API approach](https://gist.github.com/MustafaJafar/bd2a388e4a6aa3613d64a186ebb6660c?permalink_comment_id=5067094#gistcomment-5067094) remains a goal of the PR.
- [ ] Keep in mind [Fabia's usage where their AX Publish Submitter](https://github.com/ynput/ayon-core/pull/542#issuecomment-2135399068)  can submit just the publish of existing files to the farm and create e.g. a nuke dependency job that generates custom previews of the intermediate media, etc.

#### Demo scene file

The demo scene file:

[ynts_char_hero_pdg_v012.zip](https://github.com/ynput/ayon-core/files/15407437/ynts_char_hero_pdg_v012.zip)

## Testing notes:

1. Check out the branch
2. Check out the explainer video
3. Test the demo scene file
4. Comment with great ideas on how to improve
